### PR TITLE
[MBQL lib] Allow temporal bucketing of Date(Time) expressions

### LINKED ISSUE #47341: The custom datetime columns are automatically truncated by day when aggregated
### Describe the bug

The custom datetime columns in MySQL and StarRocks are automatically truncated by day when aggregated whether using functions convertTimezone() or datetimeAdd().

image

P.S.
The data types of columns are datetime in db and displayed on metabase like this:
image

### To Reproduce

-

### Expected behavior

_No response_

### Logs

_No response_

### Information about your Metabase installation

```JSON
* Browser: Chrome 128
* OS: CentOS 7
* Databases: MySQL, StarRocks
* Metabase version: v0.50.22
* Metabase hosting environment: Docker
* Metabase internal database: PostgreSQL
```

### Severity

Medium

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bshepherdson**
Fixes #47341.

### Description

The QP now supports temporal bucketing of custom expressions, but that has
not been surfaced in the UI until now.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Use a database other than H2, since it doesn't support time zones.
1. New question -> Anything with a DateTime column
1. Create a custom expression like `convertTimezone([Date Column], "America/New_York", "UTC")`
1. Break out by that expression

The temporal bucketing should work in the UI and in the query results, save properly, etc.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by bshepherdson** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/47427?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#47427** <a href="https://app.graphite.dev/github/pr/metabase/metabase/47427?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#47154** <a href="https://app.graphite.dev/github/pr/metabase/metabase/47154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bshepherdson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/metabase/lib/expression.cljc [hunk 1]
```diff
@@ -82,8 +85,13 @@
   (str s))
 
 (defmethod lib.metadata.calculation/display-name-method :expression
-  [_query _stage-number [_expression _opts expression-name] _style]
-  expression-name)
+  [_query _stage-number [_expression {:keys [temporal-unit] :as _opts} expression-name] _style]
+  (letfn [(temporal-format [display-name]
+            (lib.util/format "%s: %s" display-name (-> (name temporal-unit)
+                                                       (str/replace \- \space)
+                                                       u/capitalize-en)))]
```
Comment (lbrdnk) [inline src/metabase/lib/expression.cljc:92]: `(-> ...)` seems like a candidate for a lib function.

File: src/metabase/lib/schema/ref.cljc [hunk 2]
```diff
@@ -73,8 +73,17 @@
   (or ((some-fn :effective-type :base-type) opts)
       ::expression/type.unknown))
 
-(mbql-clause/define-tuple-mbql-clause :expression
-  #_expression-name ::common/non-blank-string)
+(mr/def ::expression.options
+  [:merge
+   ::common/options
+   [:map
+    [:temporal-unit                              {:optional true} [:ref ::temporal-bucketing/unit]]]])
```
Comment (lbrdnk) [inline src/metabase/lib/schema/ref.cljc:80]: Whitespace?

File: src/metabase/query_processor/middleware/annotate.clj [hunk 3]
```diff
@@ -138,6 +138,8 @@
 
     [:field _ (_ :guard :temporal-unit)]
     true
+    [:expression _ (_ :guard :temporal-unit)]
```
Comment (lbrdnk) [inline src/metabase/query_processor/middleware/annotate.clj:141]: Whitespace?

**Review by lbrdnk — APPROVED** [review]
Looks good! Also tested locally and all seems fine 👍

**Pull request merged**